common/interpreter/gipsy_scheme/SchemeExtensions.cpp

```diff
@@ -1046,6 +1050,312 @@ static pointer key_value_close(scheme *sc, pointer args)
+            pointer s_value = json_to_expression_r(sc, json_object_get_value(object, str));
+            s_value = sc->vptr->cons(sc, s_value, sc->NIL);
+            s_pointer = sc->vptr->cons(sc, sc->vptr->cons(sc, s_key, s_value), s_pointer);
+        }
+        return s_pointer;
+
+    case JSONString:
+        str = json_value_get_string(value);
+        return (sc->vptr->mk_string(sc, str));
+
+    case JSONBoolean:
+      if (json_value_get_boolean(value))
+          return sc->T;
+      else
+          return sc->F;
+
+    case JSONNumber:
+        num = json_value_get_number(value);
+        if (num == ((double)(int)num)) /*  check if num is integer */
+            return sc->vptr->mk_integer(sc, (long)num);
+
+        return sc->vptr->mk_real(sc, num);
+
+    case JSONNull:
+        return sc->NIL;
+
+    case JSONError:
+    default:
+        pe::ThrowIf<pe::RuntimeError>(true, "unknown error in JSON processing");
+    }
+
+    // should never reach here
+    return sc->NIL;
+}
+
+/* ----------------------------------------------------------------- */
+/* ----------------------------------------------------------------- */
+static pointer json_to_expression(scheme* sc, pointer args)
+{
+    scheme_clear_error(sc);
+
+    // ---------- message ----------
+    pointer rest = args;
+    if (! sc->vptr->is_pair(rest))
+        return scheme_return_error(sc, "missing required parameter; json string");
+
+    pointer m = sc->vptr->pair_car(rest);
+    if (! sc->vptr->is_string(m))
+        return scheme_return_error(sc, "parameter must be a string");
+
+    std::string message = strvalue(sc, m);
+
+    // Parse the contract request
+    try {
+        JsonValue parsed(json_parse_string(message.c_str()));
+        pe::ThrowIfNull(parsed.value, "failed to parse JSON expression");
+
+        return json_to_expression_r(sc, parsed);
+    }
+    catch (pdo::error::Error& e) {
+        return scheme_return_error_s(sc, format_error_message(e));
+    }
+    catch (...) {
+    }
+
+    return scheme_return_error(sc, "conversion from JSON failed");
+}
+
+/* ----------------------------------------------------------------- */
+/* ----------------------------------------------------------------- */
+static JSON_Value *expression_to_json_r(scheme *sc, pointer expr)
+{
+    JSON_Value *j_value = NULL;
+    JSON_Value *j_value_elem = NULL;
+    JSON_Array *j_array = NULL;
+    JSON_Object *j_object = NULL;
+    JSON_Status jret;
+
+    try {
+        // '(("key" value) ...)
+        if (sc->vptr->is_list(sc, expr))
+        {
+            j_value = json_value_init_object();
+            j_object = json_value_get_object(j_value);
+
+            for (pointer next = expr; next != sc->NIL; next = sc->vptr->pair_cdr(next))
+            {
+                pointer elem = sc->vptr->pair_car(next);
+                int length = sc->vptr->list_length(sc, elem);
+                pe::ThrowIf<pe::RuntimeError>(length != 2, "invalid JSON object representation");
```

Comment:
Error message is a bit missleading? It's rather an invalid s-expression not representable as JSON?

Reply:
BTW: to understand that code it is crucial to have seen the comment you had in your PR. That mapping outline should really be in the code itself as mentioned also elsewhere.

Reply:
Updated